### PR TITLE
Replaced Java 7+ references with Java 6 compatible alternatives

### PR DESCRIPTION
**Pull request by ghost**
* `java.lang.Class.getDeclaredAnnotation(Class)` -> `java.lang.Class.getDeclaredAnnotations()`
* `new java.util.ArrayList<>()` -> `com.google.common.collect.Lists.newArrayList()`
* `new java.util.HashMap<>()` -> `com.google.common.collect.Maps.newHashMap()`
* `java.util.Objects` -> `org.apache.commons.lang3.ObjectUtils`
* `java.util.Optional` -> `com.google.common.base.Optional`
* `java.util.function.Consumer` -> `com.elytradev.concrete.common.Consumer`
* `java.util.function.Predicate` -> `com.google.common.base.Predicate`
* `java.util.function.Supplier` -> `com.google.common.base.Supplier`

### CODE REVIEW

File: src/main/java/com/elytradev/concrete/block/ExpDropBehaviour.java [hunk 1]
```diff
@@ -39,21 +39,26 @@
 /**
  * An interface used to describe the behaviour for exp drops for a {@link ConcreteBlock}.
  */
-public interface ExpDropBehaviour {
+public abstract class ExpDropBehaviour {
```
Comment (exaskye) [inline src/main/java/com/elytradev/concrete/block/ExpDropBehaviour.java]: Revert changes to this class. Retrolambda supports static members and default methods.

File: src/main/java/com/elytradev/concrete/common/Consumer.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package com.elytradev.concrete.common;
+
+public interface Consumer<T> {
```
Comment (exaskye) [inline src/main/java/com/elytradev/concrete/common/Consumer.java:31]: ~~Doesn't Guava have a Consumer already?~~ It does not. Huh.

File: src/main/java/com/elytradev/concrete/config/ConcreteConfig.java [hunk 3]
```diff
@@ -66,7 +66,7 @@ public void loadConfig() {
 				String valueComment = cfgValue.comment();
 				String valueCategory = cfgValue.category();
 
-				if (Objects.equals(valueKey, "")) {
+				if (ObjectUtils.equals(valueKey, "")) {
```
Comment (exaskye) [inline src/main/java/com/elytradev/concrete/config/ConcreteConfig.java]: Prefer Guava to Apache; `Objects::equal`.

File: src/main/java/com/elytradev/concrete/inventory/ConcreteFluidTank.java [hunk 4]
```diff
@@ -76,7 +77,7 @@
  *
  */
 public class ConcreteFluidTank extends FluidTank implements IObservableFluidTank {
-    private ArrayList<Runnable> listeners = new ArrayList<>();
+    private ArrayList<Runnable> listeners = Lists.newArrayList();
```
Comment (exaskye) [inline src/main/java/com/elytradev/concrete/inventory/ConcreteFluidTank.java]: As long as you're modifying the code, swap ArrayList to List.

File: src/main/java/com/elytradev/concrete/inventory/ConcreteItemStorage.java [hunk 5]
```diff
@@ -82,8 +83,8 @@
  * 
  */
 public class ConcreteItemStorage extends ItemStackHandler implements IObservableItemHandler {
-	private ArrayList<Runnable> listeners = new ArrayList<>();
-	private ArrayList<Predicate<ItemStack>> validators = new ArrayList<>();
+	private ArrayList<Runnable> listeners = Lists.newArrayList();
+	private ArrayList<Predicate<ItemStack>> validators = Lists.newArrayList();
```
Comment (exaskye) [inline src/main/java/com/elytradev/concrete/inventory/ConcreteItemStorage.java]: Ditto

File: src/main/java/com/elytradev/concrete/inventory/ValidatedInventoryView.java [hunk 6]
```diff
@@ -41,7 +42,7 @@
 public class ValidatedInventoryView implements IInventory {
 	private final ConcreteItemStorage delegate;
 	private int[] fields = new int[0];
-	private HashMap<Integer, Supplier<Integer>> fieldDelegates = new HashMap<>();
+	private HashMap<Integer, Supplier<Integer>> fieldDelegates = Maps.newHashMap();
```
Comment (exaskye) [inline src/main/java/com/elytradev/concrete/inventory/ValidatedInventoryView.java]: HashMap -> Map

File: src/main/java/com/elytradev/concrete/inventory/gui/widget/WItemSlot.java [hunk 7]
```diff
@@ -33,14 +33,15 @@
...
+import com.google.common.collect.Lists;
 
 import net.minecraft.inventory.IInventory;
 import net.minecraft.inventory.Slot;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
 public class WItemSlot extends WWidget {
-	private ArrayList<Slot> peers = new ArrayList<>();
+	private ArrayList<Slot> peers = Lists.newArrayList();
```
Comment (exaskye) [inline src/main/java/com/elytradev/concrete/inventory/gui/widget/WItemSlot.java]: ArrayList -> List

File: src/main/java/com/elytradev/concrete/inventory/gui/widget/WPanel.java [hunk 8]
```diff
@@ -40,7 +41,7 @@
  * WContainer such that it isn't confused with Container, and we don't lose anything from the lack of abstraction.
  */
 public class WPanel extends WWidget {
-	protected ArrayList<WWidget> children = new ArrayList<>();
+	protected ArrayList<WWidget> children = Lists.newArrayList();
```
Comment (exaskye) [inline src/main/java/com/elytradev/concrete/inventory/gui/widget/WPanel.java]: Ditto

File: src/main/java/com/elytradev/concrete/network/Message.java [hunk 9]
```diff
@@ -82,15 +83,25 @@ public Message(NetworkContext ctx) {
 		
 		ClassInfo ci = classInfo.get(getClass());
 		if (ci == null) {
-			ReceivedOn ro = getClass().getDeclaredAnnotation(ReceivedOn.class);
+			ReceivedOn ro = null;
```
Comment (exaskye) [inline src/main/java/com/elytradev/concrete/network/Message.java:86]: Should probably make a getDeclaredAnnotation utility within Common and use that instead, to keep the code clean.

File: src/main/java/com/elytradev/concrete/network/NetworkContext.java [hunk 10]
```diff
@@ -100,7 +100,7 @@ public NetworkContext register(Class<? extends Message> clazz) {
 					if (f.getAnnotation(Optional.class) != null) {
 						optionalCount.add(clazz);
 					}
-					WireField<?> wf = new WireField<>(f);
+					WireField<?> wf = new WireField(f);
```
Comment (exaskye) [inline src/main/java/com/elytradev/concrete/network/NetworkContext.java]: Why're you using rawtypes here?

**Review by exaskye — CHANGES_REQUESTED** [review]

File: src/main/java/com/elytradev/concrete/common/Consumer.java (hunk 2 above)
Comment (ghost) [reply src/main/java/com/elytradev/concrete/common/Consumer.java:31]: Nope

File: src/main/java/com/elytradev/concrete/config/ConcreteConfig.java (hunk 3 above)
Comment (ghost) [reply src/main/java/com/elytradev/concrete/config/ConcreteConfig.java]: Oh, didn't know that existed

File: src/main/java/com/elytradev/concrete/block/ExpDropBehaviour.java (hunk 1 above)
Comment (ghost) [reply src/main/java/com/elytradev/concrete/block/ExpDropBehaviour.java]: Should I add `defaultMethods true` to `build.gradle` (in `retrolambda { ... }`)?

File: src/main/java/com/elytradev/concrete/network/NetworkContext.java (hunk 10 above)
Comment (ghost) [reply src/main/java/com/elytradev/concrete/network/NetworkContext.java]: At the time I made this change, I wasn't aware that diamond operators would still work with Java 6.

File: src/main/java/com/elytradev/concrete/block/ExpDropBehaviour.java (hunk 1 above)
Comment (exaskye) [reply src/main/java/com/elytradev/concrete/block/ExpDropBehaviour.java]: Yes, I thought we already had it enabled.

File: src/main/java/com/elytradev/concrete/network/NetworkContext.java (hunk 10 above)
Comment (jamierocks) [reply src/main/java/com/elytradev/concrete/network/NetworkContext.java]: They don't - they are an SE 7 feature.

Comment (exaskye) [reply src/main/java/com/elytradev/concrete/network/NetworkContext.java]: Yes, but we use Retrolambda, which supports diamond operators.

**Comment by jamierocks** [comment]
As of Minecraft 1.12, Minecraft depends on Java 8. I see no reason to make changes, we are going to shortly after revert. Using Guava in the future also isn't an option because newer versions of Guava are removing the functionality now provided by Java. Seeing as Elytra doesn't intend to appeal to the masses, and rather strives to create the most innovation, and creative mods possible - I see no reason why we would want to support Java 6, which has been out-of-support for what - 2 years now?

We've asked Mojang for it for a long time - let's not reject Java 8 now they've given it to us.

File: src/main/java/com/elytradev/concrete/network/Message.java (hunk 9 above)
Comment (ghost) [reply src/main/java/com/elytradev/concrete/network/Message.java:86]: Would Reflect be a better fit?

**Comment by ghost** [comment]
The problem of reverting back to Java 8 can be easily fixed: fork the 1.12 update from f1b19d4a6ab890f77ec0a424fd8d10f92b9970b4 (or the commit that merges #11). Of course, that means I'll have to wait until that PR is merged and then fix merge conflicts, but it's not a huge problem.

**Comment by jamierocks** [comment]
My point is really rather simple - *there is no need*. This has not been requested by anyone - *to my knowledge*. It seems Minecraft 1.13 may have some large internal changes - so Minecraft 1.12 will likely have a longer shelf-life than the previous couple versions. We may well have the next 1.7 on our hands. We should start thinking about Minecraft 1.12, not supporting players running a rather ancient version of Java.

**Comment by exaskye** [comment]
Since you seem to be comfortable rummaging through our code and fixing minor nitpicks, maybe this PR should be inverted? General cleanup and removal of Java 6 compatibility kludges? jamierocks has a point.

**Comment by ghost** [comment]
Sure, I can do that. For posterity's sake, I'll commit what changes I have for this branch and then I'll get to work on that.

**Comment by jamierocks** [comment]
You could likely start with the removal of retrolambda actually + changes ArrayList fields -> List (like previously suggested) + Lists.newArrayList()

**Comment by exaskye** [comment]
We're keeping Retrolambda, it's more safe in relation to SpecialSource.

**Pull request closed without merging**